File: 2022/Day_08/main.exs
defmodule Day_08 do
  @type forest() :: [trees()]
  @type trees() :: [tree()]
  @type tree() :: {
    height :: non_neg_integer(),
    visibility :: :N | :V,
    trees_in_sight :: non_neg_integer()
  }

  @spec parse_input(binary()) :: forest()
  def parse_input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
    |> Enum.map(fn l ->
      String.codepoints(l)
      |> Enum.map(& {String.to_integer(&1), :N, 1})
    end)
  end

  @spec calculate(forest()) :: forest()
  def calculate(input) do
    for _i <- 1..4, reduce: input do
      matrix ->
        for row <- matrix do
          head = hd(row) |> elem(0)
          [ {head, :V, 0} | (for {e, ev, ess} <- Enum.drop(row, 1), reduce: {[], head, [head]} do
            {new_row, tallest, prev} ->
              ss = Enum.reduce_while(prev, 0, fn x, acc -> if x < e, do: {:cont, acc + 1}, else: {:halt, acc + 1} end)
              new_tallest = max(e, tallest)
              { new_row ++ [{e, (if e > tallest, do: :V, else: ev), ess * ss}], new_tallest, [ e | prev ]}
          end) |> elem(0) ]
        end
        |> Enum.zip_with(& &1) |> Enum.map(&Enum.reverse/1) # rotates the matrix 90º clockwise
    end
  end

  @spec part1(forest()) :: integer()
  def part1(input) do
    Enum.reduce(input, 0, fn row, acc ->
      acc + Enum.count(row, & elem(&1,1) == :V)
    end)
  end

  @spec part2(forest()) :: integer()
  def part2(input) do
    Enum.reduce(input, 0, fn row, acc ->
      max(Enum.map(row, & elem(&1, 2)) |> Enum.max, acc)
    end)
  end

  
  def main do
    filename = Enum.at(System.argv, 0, "input.txt")
    parsed_input = parse_input(filename) |> calculate

    IO.inspect(part1(parsed_input), label: "Part 1")
    IO.inspect(part2(parsed_input), label: "Part 2")
  end
end

Day_08.main

File: 2022/Day_08/main.py
from sys import argv

with open(argv[1] if len(argv) > 1 else "input.txt") as f:
    matrix = [ [ int(x) for x in line.strip()] for line in f.readlines() ]

height = len(matrix)
width = len(matrix[0])

visible = (height + width - 2) * 2
max_visibility = 0

for j in range(1, height - 1):
    for i in range(1, width - 1):
        e = matrix[j][i]

        visible_dirs = False
        visibility = 1

        for dir in [
          matrix[j][i-1::-1], # left
          matrix[j][i+1:],  # right
          map(lambda row: row[i], matrix[j-1::-1]), # top
          map(lambda row: row[i], matrix[j+1:])]: # bottom
            visibility_dir = 0
            visible_dir = True

            for a in dir:
                visibility_dir += 1
                if a >= e:
                    visible_dir = False
                    break

            visible_dirs = visible_dirs or visible_dir
            visibility *= visibility_dir
                
        if visible_dirs: visible += 1

        max_visibility = max(max_visibility, visibility)

print(f"Part 1: {visible}")
print(f"Part 2: {max_visibility}")